README.md
# Repsheet [![Build Status](https://secure.travis-ci.org/repsheet/repsheet.png)](http://travis-ci.org/repsheet/repsheet?branch=master)

> 
> **noun**  *Slang.*  
> A record kept by web site authorities of a machine's wrongdoings.
>
> **noun**  *Slang.*  
> A necessary tool in your web defense arsenal.
>
> **Origin:**  
> [2013]

Repsheet is a collection of tools to help improve awareness of robots and bad actors visiting your web applications. It is primarily an Apache web server module with a couple of addons that help organize and aggregate the data that it collects.

## What problem does it solve?

Repsheet attempts to solve the problem of automated defenses against robots and noisy attackers. It collects and records activity of the IP addresses that access your web sites and helps determine if they are misbehaving. If they are, they are put on the Repsheet. Once there, you can choose to outright deny them access to your site or warn your downstream applications that the requestor is a known bad actor and that the request should be handled differently. Essentially, it is a local reputation engine.

## How does it work?

Repsheet works by inspecting requests as they come into the web server and storing that activity in Redis. It does this in two ways. 

First, when a request comes in, the IP is checked to see if it has been flagged by Repsheet. If it has, it acts on that information based on the configuration set. If `RepsheetAction` is set to `Block`, Repsheet instructs Apache to return a 403. If `RepsheetAction` is set to `Notify`, then Repsheet will log to the Apache logs that the IP was found on the repsheet but no action was taken. Repsheet will then add the header `X-Repsheet: true` to the request to make downstream applications aware that the request is from a known bad source. Next, if the `RepsheetRecorder` directive is set to `On`, the the following details are recorded in Redis list using the key `<ip>:requests`

* Timestamp
* User Agent
* HTTP Method
* URI
* Arguments

These are recorded for every request the IP makes.

Second, if the `RepsheetFilter` directive is set to `On`, Repsheet looks at [ModSecurity](http://modsecurity.org) to see if the request triggered any rules. If so, [ModSecurity](http://modsecurity.org) places details into the request headers. Repsheet reads these headers and records information about which rules were triggered and stores them in Redis in the following fashion:

```
sadd <ruleid> <ip>:detected
incr <ip>:<rule>:count
set <ip>:repsheet true
```

In other words, each IP address that triggers a ModSecurity is put on the Repsheet and each rule that has been triggered is recorded as well as how many times that rule has been triggered. Finally, The IP address is flagged as being a repsheet offender.

The high level idea is captured in this drawing:

![Repsheet](https://raw.github.com/abedra/repsheet/master/doc/Repsheet.png)

## What are the benefits?

Repsheet provides some much needed intelligence and visualization around who is acting against your web site(s) and how they might be attacking. All of this data is available when the infrastructure is setup properly, but it isn't correlated or aggregated in any meaningful way. On top of these benefits, it allows for easy addition of an IP to the repsheet by just adding it manually in Redis. This means that you can plug in any activity processing systems you want and have them feed back into the Redis instance that Repsheet reads from. This makes the solution as flexible as you need it to be. It also leaves space for statistical analysis of traffic patterns to run in the background on this data and classify/detect anomalies that can be fed back into Repsheet.

## What are the downsides?

In short, more work. A Redis instance has to be spun up that can handle the data coming in to your site. If you get a lot of traffic, your Redis instance is going to grow. You can configure how many requests are stored for an IP address to limit a single IP address from blowing up your Redis database, but you still have to account for all the IPs that come into contact and how frequently they do so. 

Along with this, there is the potential for both [ModSecurity](http://modsecurity.org) and Repsheet to slow down your requests. The [ModSecurity](http://modsecurity.org) slow downs are mostly based on the rules you configure, but you should do some profiling to ensure that [ModSecurity](http://modsecurity.org) isn't causing to large of a delay. Since Repsheet contacts an external service during the request lifecycle in a blocking fashion, the request is slowed down for the duration of that external service call. Fortunately this is configurable in `RepsheetRedisTimeout`. You should keep this as low as possible, and the timeout value is always in milliseconds. This should be profiled as well to ensure that it doesn't cause unnecessary overhead.

## Setup

This module requires [hiredis](https://github.com/redis/hiredis), [apxs](http://httpd.apache.org/docs/2.2/programs/apxs.html), and [pcre](http://www.pcre.org/), and optionally [geoip](http://www.maxmind.com/en/geolocation_landing) (if you want GeoIP support) to be installed. If you want to run the integration tests, you will need to have [Ruby](http://www.ruby-lang.org/en/) and [RubyGems](http://rubygems.org/) installed. The integration tests use [Bundler](http://gembundler.com/), so you need to have that installed as well. The Ruby based programs have all been tested using Ruby 1.9.3.

If you are on OS X, you might have to run the following to make the compile tool-chain work properly with apxs:

```
ln -s /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/ /Applications/Xcode.app/Contents/Developer/Toolchains/OSX10.8.xctoolchain
```

#### Installation

Compilation is done via `apxs`. This is one of the simpler ways to deal with compilation/installation/activation of a module. Repsheet uses autotools to generate `configure` scripts and make files. If you want to build from source you will need to have the [check](http://check.sourceforge.net/) library installed as well as the dependencies listed above.

```
autogen.sh
./configure
make
sudo make install
```

To activate and configure repsheet you will need to set some directives. The following list explains what each directive is and what is does.

* `RepsheetEnabled <On|Off>` - Determines if Repsheet will do any processing
* `RepsheetRecorder <On|Off>` - Determines if request information will be stored
* `RepsheetFilter <On|Off>` - Determines if Repsheet will look for ModSecurity information
* `RepsheetGeoIP <On|Off>` - Determines if Repsheet will look at GeoIP information
* `RepsheetProxyHeaders <On|Off>` - Determines if Repsheet will look for the X-Forwarded-For header to determine remote ip
* `RepsheetAction <Notify|Block>` - Determines the action to take if an IP is found on the repsheet. `Notify` will send a header downstream and `Block` will return a `403`
* `RepsheetPrefix <prefix>` - Sets the logger prefix. This will precede any repsheet apache log lines
* `RepsheetRedisTimeout n` - Sets the time (in milliseconds) before the attempt to connect to redis will timeout and fail
* `RepsheetRedisHost <host>` - Sets the host for the Redis connection
* `RepsheetRedisTimeout <port>` - Sets the port for the Redis connection
* `RepsheetRedisMaxLength <max>` - Number of recorded requests a single IP can have before it is trimmed in Redis
* `RepsheetRedisExpiry <hours>` - Number of hours of inactivity before an entry expires

Here's a complete example:

```
<IfModule repsheet_module>
  RepsheetEnabled On
  RepsheetRecorder On
  RepsheetFilter On
  RepsheetGeoIP On
  RepsheetProxyHeaders On
  RepsheetAction Notify
  RepsheetPrefix [repsheet]
  RepsheetRedisTimeout 5
  RepsheetRedisHost localhost
  RepsheetRedisPort 6379
  RepsheetRedisMaxLength 1000
  RepsheetRedisExpiry 24
</IfModule>
```

##### ModSecurity Settings

In order to take full advantage of Repsheet, you need to have the proper ModSecurity rules setup. This repository packages a working set of rules as an example. They are the  [OWASP Core Rules](https://github.com/SpiderLabs/owasp-modsecurity-crs) base rules plus the optional rule set that perfoms header tagging. You can and should tune these rules so that they work properly for your enviornment, but remember to have the header tagging rules in place or Repsheet will not be able to see what ModSecurity has alerted on.

## Running the Unit Tests

Repsheet has a set of C based unit tests for the core logic. You can run them with `make check` after you have setup your build environment as described above.


## Running the Integration Tests

Repsheet comes with a basic set of integration tests to ensure that things are working. In order to run these, run the following

```sh
script/bootstrap
autogen.sh
configure
make
build/bin/apxs -i -a src/mod_security.la
bundle install
rake
```

The `script/bootstrap` task will take some time. It downloads and compiles Apache and ModSecurity, and then compiles and installs Repsheet and configures everything to work together. Running `rake` launches some curl based tests that hit the site and exercise Repsheet, then test that everything was recorded properly in Redis.


debian/changelog
libapache2-mod-repsheet (0.9-1) unstable-security; urgency=low

  * Fixes an issue where Repsheet held Redis connections open when blocking actors

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Wed, 24 Jul 2013 12:12:18 -0500

libapache2-mod-repsheet (0.8-1) unstable-security; urgency=low

  * Store triggered rules in a sorted set

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Mon, 22 Jul 2013 10:04:05 -0500

libapache2-mod-repsheet (0.7-1) unstable-security; urgency=low

  * Use the same regex for calculating ModSecurity matches

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Wed, 03 Jul 2013 11:05:05 -0500

libapache2-mod-repsheet (0.6-1) unstable-security; urgency=low

  * Fix ModSecurity header parsing regex to be more restrictive

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Wed, 26 Jun 2013 20:35:22 +0000

libapache2-mod-repsheet (0.5-1) unstable-security; urgency=low

  * Fixes issue #20 (Expire all keys when setting expiry)

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Mon, 24 Jun 2013 14:07:30 +0000

libapache2-mod-repsheet (0.4-1) unstable-security; urgency=low

  * Adding version number to logs on startup

 -- Aaron Bedra <aaron@aaronbedra.com>  Fri, 14 Jun 2013 23:16:12 -0400

libapache2-mod-repsheet (0.3-1) unstable-security; urgency=low

  * Adding configurable record and repsheet key expiry

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Thu, 13 Jun 2013 18:14:18 +0000

libapache2-mod-repsheet (0.2-2) unstable-security; urgency=low

  * Fixes install directory package issue

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Wed, 05 Jun 2013 21:24:39 +0000

libapache2-mod-repsheet (0.2-1) unstable-security; urgency=low

  * Updating packaging to reflect new autotools project structure

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Wed, 05 Jun 2013 16:37:03 +0000

libapache2-mod-repsheet (0.1-1) unstable-security; urgency=low

  * Initial release.

 -- Aaron Bedra <aaron.bedra@braintreepayments.com>  Tue, 30 Apr 2013 19:18:45 +0000


src/mod_repsheet.c
/*
  Copyright 2013 Aaron Bedra

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

#include <stdio.h>

#include "http_core.h"
#include "http_protocol.h"
#include "http_request.h"
#include "http_config.h"
#include "http_log.h"

#include "hiredis/hiredis.h"

#include "proxy.h"
#include "mod_security.h"
#include "repsheet.h"

#define REPSHEET_VERSION "0.9"

typedef struct {
  int repsheet_enabled;
  int recorder_enabled;
  int filter_enabled;
  int proxy_headers_enabled;
  int geoip_enabled;
  int action;
  const char *prefix;

  const char *redis_host;
  int redis_port;
  int redis_timeout;
  int redis_max_length;
  int redis_expiry;
} repsheet_config;
static repsheet_config config;

const char *repsheet_set_enabled(cmd_parms *cmd, void *cfg, const char *arg)
{
  if (!strcasecmp(arg, "on")) {
    config.repsheet_enabled = 1;
  } else {
    config.repsheet_enabled = 0;
  }
  return NULL;
}

const char *repsheet_set_recorder_enabled(cmd_parms *cmd, void *cfg, const char *arg)
{
  if (!strcasecmp(arg, "on")) {
    config.recorder_enabled = 1;
  } else {
    config.recorder_enabled = 0;
  }
  return NULL;
}

const char *repsheet_set_filter_enabled(cmd_parms *cmd, void *cfg, const char *arg)
{
  if (!strcasecmp(arg, "on")) {
    config.filter_enabled = 1;
  } else {
    config.filter_enabled = 0;
  }
  return NULL;
}

const char *repsheet_set_geoip_enabled(cmd_parms *cmd, void *cfg, const char *arg)
{
  if (!strcasecmp(arg, "on")) {
    config.geoip_enabled = 1;
  } else {
    config.geoip_enabled = 0;
  }
  return NULL;
}

const char *repsheet_set_proxy_headers_enabled(cmd_parms *cmd, void *cfg, const char *arg)
{
  if (!strcasecmp(arg, "on")) {
    config.proxy_headers_enabled = 1;
  } else {
    config.proxy_headers_enabled = 0;
  }
  return NULL;
}

const char *repsheet_set_timeout(cmd_parms *cmd, void *cfg, const char *arg)
{
  config.redis_timeout = atoi(arg) * 1000;
  return NULL;
}

const char *repsheet_set_host(cmd_parms *cmd, void *cfg, const char *arg)
{
  config.redis_host = arg;
  return NULL;
}

const char *repsheet_set_port(cmd_parms *cmd, void *cfg, const char *arg)
{
  config.redis_port = atoi(arg);
  return NULL;
}

const char *repsheet_set_length(cmd_parms *cmd, void *cfg, const char *arg)
{
  config.redis_max_length = atoi(arg);
  return NULL;
}

const char *repsheet_set_expiry(cmd_parms *cmd, void *cfg, const char *arg)
{
  config.redis_expiry = atoi(arg) * 60 * 60;
  return NULL;
}

const char *repsheet_set_prefix(cmd_parms *cmd, void *cfg, const char *arg)
{
  config.prefix = arg;
  return NULL;
}

const char *repsheet_set_action(cmd_parms *cmd, void *cfg, const char *arg)
{
  if (strcmp(arg, "Notify") == 0) {
    config.action = NOTIFY;
  } else if (strcmp(arg, "Block") == 0) {
    config.action = BLOCK;
  } else {
    config.action = ALLOW;
  }
  return NULL;
}

static const command_rec repsheet_directives[] =
  {
    AP_INIT_TAKE1("repsheetEnabled",         repsheet_set_enabled,               NULL, RSRC_CONF, "Enable or disable mod_repsheet"),
    AP_INIT_TAKE1("repsheetRecorder",        repsheet_set_recorder_enabled,      NULL, RSRC_CONF, "Enable or disable repsheet recorder"),
    AP_INIT_TAKE1("repsheetFilter",          repsheet_set_filter_enabled,        NULL, RSRC_CONF, "Enable or disable repsheet ModSecurity filter"),
    AP_INIT_TAKE1("repsheetGeoIP",           repsheet_set_geoip_enabled,         NULL, RSRC_CONF, "Enable or disable repsheet GeoIP filter"),
    AP_INIT_TAKE1("repsheetProxyHeaders",    repsheet_set_proxy_headers_enabled, NULL, RSRC_CONF, "Enable or disable proxy header scanning"),
    AP_INIT_TAKE1("repsheetAction",          repsheet_set_action,                NULL, RSRC_CONF, "Set the action"),
    AP_INIT_TAKE1("repsheetPrefix",          repsheet_set_prefix,                NULL, RSRC_CONF, "Set the log prefix"),
    AP_INIT_TAKE1("repsheetRedisTimeout",    repsheet_set_timeout,               NULL, RSRC_CONF, "Set the Redis timeout"),
    AP_INIT_TAKE1("repsheetRedisHost",       repsheet_set_host,                  NULL, RSRC_CONF, "Set the Redis host"),
    AP_INIT_TAKE1("repsheetRedisPort",       repsheet_set_port,                  NULL, RSRC_CONF, "Set the Redis port"),
    AP_INIT_TAKE1("repsheetRedisMaxLength",  repsheet_set_length,                NULL, RSRC_CONF, "Last n requests kept per IP"),
    AP_INIT_TAKE1("repsheetRedisExpiry",     repsheet_set_expiry,                NULL, RSRC_CONF, "Number of hours before records expire"),
    { NULL }
  };

static redisContext *get_redis_context(request_rec *r)
{
  redisContext *context;
  struct timeval timeout = {0, (config.redis_timeout > 0) ? config.redis_timeout : 10000};

  context = redisConnectWithTimeout((char*) config.redis_host, config.redis_port, timeout);
  if (context == NULL || context->err) {
    if (context) {
      ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, "%s Redis Connection Error: %s", config.prefix, context->errstr);
      redisFree(context);
    } else {
      ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, "%s Connection Error: can't allocate redis context", config.prefix);
    }
    return NULL;
  } else {
    return context;
  }
}

static char *remote_address(request_rec *r)
{
  if (config.proxy_headers_enabled) {
    char *address = process_headers((char*)apr_table_get(r->headers_in, "X-Forwarded-For"));

    if (address == NULL) {
      return r->connection->remote_ip;
    }

    return address;
  } else {
    return r->connection->remote_ip;
  }
}

static void record(redisContext *context, request_rec *r)
{
  apr_time_exp_t start;
  char           timestamp[50];
  char           *ip = remote_address(r);

  apr_time_exp_gmt(&start, r->request_time);
  sprintf(timestamp, "%d/%d/%d %d:%d:%d.%d", (start.tm_mon + 1), start.tm_mday, (1900 + start.tm_year), start.tm_hour, start.tm_min, start.tm_sec, start.tm_usec);

  repsheet_record(context, timestamp, apr_table_get(r->headers_in, "User-Agent"), r->method, r->uri, r->args, ip, config.redis_max_length, config.redis_expiry);
}

static void process_waf_events(redisContext *context, request_rec *r, char *waf_events)
{
  int i, m;
  char **events;

  m = matches(waf_events);

  if (m > 0) {
    events = malloc(m * sizeof(char*));
    for(i = 0; i < m; i++) {
      events[i] = malloc(i * sizeof(char));
    }

    process_mod_security_headers(waf_events, events);

    char *ip = remote_address(r);

    for(i = 0; i < m; i++) {
      freeReplyObject(redisCommand(context, "ZINCRBY %s:detected 1 %s", ip, events[i]));
      freeReplyObject(redisCommand(context, "SET %s:repsheet true", ip));
      if (config.redis_expiry > 0) {
        freeReplyObject(redisCommand(context, "EXPIRE %s:detected %d", ip, config.redis_expiry));
        freeReplyObject(redisCommand(context, "EXPIRE %s:repsheet %d", ip, config.redis_expiry));
      }
    }
    free(events);
  }
}

static int repsheet_lookup(request_rec *r)
{
  if (!config.repsheet_enabled) {
    return DECLINED;
  }

  redisContext *context = get_redis_context(r);

  if (context == NULL) {
    return DECLINED;
  }

  int action;
  char *ip = remote_address(r);

  action = repsheet_ip_lookup(context, ip);
  if (action) {
    if (action == BLOCK) {
      ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, "%s %s was blocked by the repsheet", config.prefix, ip);
      redisFree(context);
      return HTTP_FORBIDDEN;
    } else if (action == NOTIFY) {
      if (config.action == BLOCK) {
        ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, "%s %s was blocked by the repsheet", config.prefix, ip);
        redisFree(context);
        return HTTP_FORBIDDEN;
      } else {
        ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, "%s IP Address %s was found on the repsheet. No action taken", config.prefix, ip);
        apr_table_set(r->headers_in, "X-Repsheet", "true");
      }
    }
  }

  redisFree(context);

  return DECLINED;
}

static int repsheet_recorder(request_rec *r)
{
  if (!config.repsheet_enabled || !config.recorder_enabled || !ap_is_initial_req(r)) {
    return DECLINED;
  }

  redisContext *context = get_redis_context(r);

  if (context == NULL) {
    return DECLINED;
  }

  record(context, r);

  redisFree(context);

  return DECLINED;
}

static int repsheet_geoip_filter(request_rec *r)
{
  if (!config.repsheet_enabled || !config.geoip_enabled) {
    return DECLINED;
  }

  const char* country = apr_table_get(r->headers_in, "GEOIP_COUNTRY_CODE");

  if (country == NULL) {
    return DECLINED;
  }

  redisContext *context = get_redis_context(r);

  if (context == NULL) {
    return DECLINED;
  }

  int action;
  char *ip = remote_address(r);

  action = repsheet_geoip_lookup(context, country);

  if (action) {
    if (action == BLOCK) {
      ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, "%s %s was blocked by the repsheet", config.prefix, ip);
      return HTTP_FORBIDDEN;
    } else if (action == NOTIFY) {
      if (config.action == BLOCK) {
        ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, "%s %s was blocked by the repsheet", config.prefix, ip);
        return HTTP_FORBIDDEN;
      } else {
        ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, "%s IP Address %s was on the geoip suspicious country list. No action taken", config.prefix, ip);
        apr_table_set(r->headers_in, "X-Repsheet", "true");
      }
    }

    freeReplyObject(redisCommand(context, "SET %s:repsheet true", ip));
    if (config.redis_expiry > 0) {
      freeReplyObject(redisCommand(context, "EXPIRE %s:repsheet %d", ip, config.redis_expiry));
    }
  }

  redisFree(context);

  return DECLINED;
}

static int repsheet_mod_security_filter(request_rec *r)
{
  if (!config.repsheet_enabled || !config.filter_enabled || !ap_is_initial_req(r)) {
    return DECLINED;
  }

  char *waf_events = (char *)apr_table_get(r->headers_in, "X-WAF-Events");

  if (!waf_events) {
    return DECLINED;
  }

  redisContext *context = get_redis_context(r);

  if (context == NULL) {
    return DECLINED;
  }

  process_waf_events(context, r, waf_events);

  redisFree(context);

  return DECLINED;
}

static int hook_post_config(apr_pool_t *mp, apr_pool_t *mp_log, apr_pool_t *mp_temp, server_rec *s) {
  ap_log_error(APLOG_MARK, APLOG_NOTICE | APLOG_NOERRNO, 0, s, "Repsheet Version %s Enabled", REPSHEET_VERSION);
  return OK;
}

static void register_hooks(apr_pool_t *pool)
{
  ap_hook_post_config(hook_post_config, NULL, NULL, APR_HOOK_REALLY_LAST);
  ap_hook_post_read_request(repsheet_lookup, NULL, NULL, APR_HOOK_LAST);
  ap_hook_post_read_request(repsheet_recorder, NULL, NULL, APR_HOOK_LAST);
  ap_hook_post_read_request(repsheet_geoip_filter, NULL, NULL, APR_HOOK_LAST);
  ap_hook_fixups(repsheet_mod_security_filter, NULL, NULL, APR_HOOK_REALLY_LAST);
}

module AP_MODULE_DECLARE_DATA repsheet_module =
  {
    STANDARD20_MODULE_STUFF,
    NULL,                /* Per-directory configuration handler */
    NULL,                /* Merge handler for per-directory configurations */
    NULL,                /* Per-server configuration handler */
    NULL,                /* Merge handler for per-server configurations */
    repsheet_directives, /* Any directives we may have for httpd */
    register_hooks       /* Our hook registering function */
  };
